src/realm/db.hpp

```diff
@@ -686,31 +698,52 @@ class Transaction : public Group {
     }
     // request full synchronization to stable storage for all writes done since
     // last sync. The write mutex is released after full synchronization.
-    void async_request_sync_to_storage(std::function<void()> when_synchronized);
+    void async_request_sync_to_storage(std::function<void()> when_synchronized = nullptr);
     // release the write lock without any sync to disk
     void async_release_write_lock()
     {
-        REALM_ASSERT(m_async_stage == AsyncState::HasLock);
-        m_async_stage = AsyncState::Idle;
-        get_db()->async_end_write();
+        std::unique_lock<std::mutex> lck(mtx);
+        if (m_async_stage == AsyncState::HasLock) {
+            m_async_stage = AsyncState::Idle;
+            get_db()->async_end_write();
+        }
     }
 
     // true if sync to disk has been requested
-    bool is_synchronizing() const noexcept
+    bool is_synchronizing() noexcept
     {
+        std::unique_lock<std::mutex> lck(mtx);
         return m_async_stage == AsyncState::Syncing;
     }
 
-    // methods for use with async interface for cont. transactions
-    DB::ReadLockInfo grab_read_lock()
+    void wait_for_write_lock()
     {
-        DB::ReadLockInfo rli;
-        get_db()->grab_read_lock(rli, VersionID());
-        return rli;
+        std::unique_lock<std::mutex> lck(mtx);
+        if (m_async_stage == Transaction::AsyncState::Requesting) {
+            waiting_for_write_lock = true;
+            cv.wait(lck);
```

Comment:
this call may wake up spuriously, so need to be inside a checking loop

Reply:
Done